File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.*;
import java.util.*;

class Main {
    static int n;
    static int m;
    static int[][] map;
    static int ans = 0;
    static boolean[][] visit;
    static int[]dx = {1, -1, 0, 0};
    static int[]dy = {0, 0, 1, -1};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        map = new int[n][m];
        visit = new boolean[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                visit[i][j] = true;
                dfs(i , j, map[i][j], 1);
                visit[i][j] = false;
            }
        }
        System.out.println(ans);
    }

    private static void dfs(int x, int y, int sum, int cnt) {
        if(cnt == 4) {
            ans = Math.max(ans, sum);
            return;
        }

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if(!Check(nx, ny)) continue;

            if(!visit[nx][ny]) {

                if(cnt == 2) {
                    // ㅗ 모양을 위한 dfs
                    visit[nx][ny] = true;
                    dfs(x, y, sum + map[nx][ny], cnt + 1);
                    visit[nx][ny] = false;
                }

                visit[nx][ny] = true;
                dfs(nx, ny, sum + map[nx][ny], cnt + 1);
                visit[nx][ny] = false;
            }
        }
    }

    private static boolean Check(int nx, int ny) {
        return 0 <= nx && nx < n && 0 <= ny && ny < m;
    }
}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
from sys import stdin as input
n,m = map(int,input.readline().split())
arr = [list(map(int,input.readline().split())) for _ in range(n)]
# i,j에서 4칸 이동했을 경우의 최댓값만 구하기
directions = [(0,1),(1,0),(-1,0),(0,-1)]
ans = 0
visit = [[False] * m for _ in range(n)]
def check(x,y):
    return 0 <= x < n and 0 <= y < m
# 'ㅗ' 모양은? -> 탐색한 척하고 다시 탐색
def dfs(x,y,s,cnt):
    global ans,arr,visit
    if cnt == 4:
        ans = max(ans,s)
        return
    
    for d in directions:
        nx,ny = x + d[0], y + d[1]
        if check(nx,ny) and not visit[nx][ny]:
            if cnt == 2:
                visit[nx][ny] = True
                # 다시 제자리에서 dfs
                dfs(x,y,s+arr[nx][ny],cnt+1)
                visit[nx][ny] = False
            visit[nx][ny] = True
            dfs(nx,ny,s+arr[nx][ny],cnt+1)
            visit[nx][ny] = False
for i in range(n):
    for j in range(m):
        visit[i][j] = True
        dfs(i,j,arr[i][j],1)
        visit[i][j] = False
print(ans)   